osu.Game/Screens/Edit/Compose/Components/Timeline/TimelineControlPointDisplay.cs

```diff
@@ -23,34 +32,73 @@ protected override void LoadBeatmap(EditorBeatmap beatmap)
 
             controlPointGroups.UnbindAll();
             controlPointGroups.BindTo(beatmap.ControlPointInfo.Groups);
-            controlPointGroups.BindCollectionChanged((_, args) =>
+            controlPointGroups.BindCollectionChanged((_, _) =>
             {
-                switch (args.Action)
-                {
-                    case NotifyCollectionChangedAction.Reset:
-                        Clear();
-                        break;
+                invalidateGroups();
+            }, true);
+        }
 
-                    case NotifyCollectionChangedAction.Add:
-                        Debug.Assert(args.NewItems != null);
+        protected override void Update()
+        {
+            base.Update();
 
-                        foreach (var group in args.NewItems.OfType<ControlPointGroup>())
-                            Add(new TimelineControlPointGroup(group));
-                        break;
+            if (timeline == null || DrawWidth <= 0) return;
+
+            (float, float) newRange = (
+                (ToLocalSpace(timeline.ScreenSpaceDrawQuad.TopLeft).X - TopPointPiece.WIDTH) / DrawWidth * Content.RelativeChildSize.X,
+                (ToLocalSpace(timeline.ScreenSpaceDrawQuad.TopRight).X) / DrawWidth * Content.RelativeChildSize.X);
+
+            if (visibleRange != newRange)
+            {
+                visibleRange = newRange;
+                invalidateGroups();
+            }
+
+            if (!groupCache.IsValid)
+                recreateDrawableGroups();
+        }
+
+        private void invalidateGroups() => groupCache.Invalidate();
 
-                    case NotifyCollectionChangedAction.Remove:
-                        Debug.Assert(args.OldItems != null);
+        private void recreateDrawableGroups()
+        {
+            // Remove groups outside the visible range
+            for (int i = Count - 1; i >= 0; i--)
+            {
+                var g = Children[i];
+
+                if (!shouldBeVisible(g.Group))
+                    g.Expire();
+            }
```

Comment:
> All expiration does is set `LifetimeEnd`

Which will make `ShouldBeAlive` false and since [`RemoveWhenNotAlive`](https://github.com/ppy/osu-framework/blob/b6d74e91d813fc21cc7cc7c6dd347502a933b53d/osu.Framework/Graphics/Drawable.cs#L1462) isn't overriden - it will be removed.
... but later in a separate pass, so you are right.